### PR TITLE
ec2_ami: Add support for params BootMode, TpmSupport, UefiData

### LINKED ISSUE #944: Boot mode and TPM support for AMI creation
### Summary

I would like to use ansible to create an AMI with UEFI secure boot and TPM support. In the [ec2_ami module](https://docs.ansible.com/ansible/latest/collections/amazon/aws/ec2_ami_module.html#ansible-collections-amazon-aws-ec2-ami-module), there are currently no parameters to do so. As UEFI and TPM support for Linux AMIs is new for AWS, I think it would be a useful new feature to add parameters for boot mode and tpm support. [Amazon EC2 Now Supports NitroTPM and UEFI Secure Boot](https://aws.amazon.com/blogs/aws/amazon-ec2-now-supports-nitrotpm-and-uefi-secure-boot/)

### Issue Type

Feature Idea

### Component Name

ec2_ami

### Additional Information

```yaml (paste below)
 - name: Create AMI
    amazon.aws.ec2_ami:
         name: test
         state: present
         architecture: x86_64
         virtualization_type: hvm
         root_device_name: /dev/sda1
         device_mapping:
                 - device_name: /dev/sda1
                    snapshot_id: "{{ snapshot_id }}"
          wait: yes
          region: us-east-1
          boot_mode: uefi
          uefi_data: data_file.bin
          tpm_support: v2.0  
                                       
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mandar242**
##### SUMMARY
Depends-On: https://github.com/ansible-collections/amazon.aws/pull/1066

Added support for params BootMode, TpmSupport, UefiData in `ec2_ami`.

Fixes #944 
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_ami
##### ADDITIONAL INFORMATION

Example playbook

```
- name: abc
  hosts: localhost
  gather_facts: false
  tasks:
    - name: AMI Creation with boot_mode and tpm_support
      amazon.aws.ec2_ami:
        name: ami-create-test_legacy-bios
        state: present
        architecture: x86_64
        virtualization_type: hvm
        root_device_name: /dev/sda1
        device_mapping:
          - device_name: /dev/sda1
            snapshot_id: snap-xxxxxxxxx
        wait: yes
        region: us-east-2
        boot_mode: legacy-bios
        tpm_support: v2.0
        tags:
          name: ami-create-test

```

### CODE REVIEW

**Comment by mandar242** [comment]
recheck

**Comment by mandar242** [comment]
recheck

**Comment by mandar242** [comment]
recheck

File: plugins/modules/ec2_ami.py [hunk 1]
```diff
@@ -143,6 +143,26 @@
...
+      - The boot mode of the AMI.
+      - See the AWS documentation for more detail U(https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ami-boot.html).
+    type: str
+    choices: ['legacy-bios', 'uefi']
+  tpm_support:
+    description:
+      - Set to v2.0 to enable Trusted Platform Module (TPM) support.
+      - If the image is configured for NitroTPM support, the value is v2.0 .
+      - Requires I(boot_mode) to be set to 'uefi'.
+      - Requires minimum botocore version 1.26.0.
```
Comment (goneri) [inline plugins/modules/ec2_ami.py:157]: I would also mention that it requires an instance type that is compatible with Nitro.

File: plugins/modules/ec2_ami.py [hunk 2]
```diff
@@ -507,6 +546,14 @@ def create_image(module, connection):
                 params['KernelId'] = kernel_id
             if root_device_name:
                 params['RootDeviceName'] = root_device_name
+            if boot_mode:
+                params['BootMode'] = boot_mode
+            if tpm_support:
+                if boot_mode != 'uefi':
```
Comment (goneri) [inline plugins/modules/ec2_ami.py:552]: You can do the check early around line 480.

File: tests/integration/targets/ec2_ami/tasks/main.yml [hunk 3]
```diff
@@ -628,6 +628,50 @@
...
+      ignore_errors: true
+
+    - name: assert that new ami has been created with desired options
+      assert:
+        that:
+          - "result.changed"
+          - "result.image_id.startswith('ami-')"
+          - ami_facts_result_boot_tpm.images[0].image_id | length != 0
+          - ami_facts_result_boot_tpm.images[0].boot_mode == 'uefi'
+          - ami_facts_result_boot_tpm.images[0].tpm_support == 'v2.0'
```
Comment (goneri) [inline tests/integration/targets/ec2_ami/tasks/main.yml:674]: Can you also test `uefi_data`?

File: plugins/modules/ec2_ami.py [hunk 4]
```diff
@@ -214,6 +234,22 @@
...
+    architecture: x86_64
+    virtualization_type: hvm
+    root_device_name: /dev/sda1
+    device_mapping:
+        - device_name: /dev/sda1
+          snapshot_id: "{{ snapshot_id }}"
+    wait: yes
+    region: us-east-1
+    boot_mode: uefi
+    uefi_data: data_file.bin
```
Comment (goneri) [inline plugins/modules/ec2_ami.py:250]: UefiData should be a base64 string, so something like ` "{{ lookup('file', '/etc/fstab')|b64encode }}"`.

**Review by goneri — CHANGES_REQUESTED** [review]

File: tests/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:674]: trying to figure out what a valid `uefi_data_file` would look like, currently boto is throwing error on below task
```
    - name: AMI Creation with boot_mode and tpm_support
      amazon.aws.ec2_ami:
        name: newtest-uefi-file-data
        state: present
        architecture: x86_64
        virtualization_type: hvm
        root_device_name: /dev/sda1
        device_mapping:
          - device_name: /dev/sda1
            snapshot_id: snap-0f00cba7123456
        wait: yes
        region: us-east-2
        boot_mode: uefi
        uefi_data: "{{ lookup('file', 'uefi_data_file') | b64encode }}"
        tpm_support: v2.0
        tags:
          name: test-ami
```

```
botocore.exceptions.ClientError: An error occurred (InvalidParameterValue)
when calling the RegisterImage operation: Invalid uefiData.
```

File: plugins/modules/ec2_ami.py [hunk 5]
```diff
@@ -247,7 +248,7 @@
     wait: yes
     region: us-east-1
     boot_mode: uefi
-    uefi_data: data_file.bin
+    uefi_data: "{{ lookup('file', '/etc/fstab')| b64encode }}"
```
Comment (goneri) [inline plugins/modules/ec2_ami.py:251]: Well, the `/etc/fstab` was a random example :-).

**Comment by mandar242** [comment]
recheck

File: tests/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (goneri) [reply tests/integration/targets/ec2_ami/tasks/main.yml:674]: I would suggest to use `get_instance_uefi_data()` https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.get_instance_uefi_data to get a working uefidata file and you can send it back later in the tests.

**Comment by goneri** [comment]
regate

**Comment by goneri** [comment]
regate

File: tests/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:676]: tried that but turns out it throws `unauthorized` error for the operator `GetInstanceUefiData`.
Would need to open a PR to get the permissions added to terminator repository.

Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:688]: @goneri opened PR to add required permission.
https://github.com/mattclay/aws-terminator/pull/229

Task to get uefi data in integration tests: https://github.com/ansible-collections/amazon.aws/pull/1037/files#diff-ceba09bf3546fc4c878b0ae63333fa702ef39a1c13fb206f78333031fd67dc07R77-R86

**Comment by goneri** [comment]
recheck

**Comment by gravesm** [comment]
recheck

File: tests/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:695]: the permission issue has been resolved, currently trying to solve error
```
An error occurred (InvalidInstanceID.NotFound) when calling the GetInstanceUefiData operation: 
Unknown virtIds for owner 'xxxxxxx': [ i-xxxxxxxxxx ]"
```

Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:695]: @goneri, it appears that there are currently no provided `AMIs` with `uefi boot mode` support for `linux`, also none of our currently [supported instance types](https://github.com/mattclay/aws-terminator/blob/master/aws/policy/compute.yaml#L24-L29) for integration tests support `uefi boot mode`.
I can test the code for `UEFI boot mode` and `UEFI data` manually but not sure if it's possible to add to `integration tests`.

Relevant links:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/launch-instance-with-uefi-sb.html
https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/enable-nitrotpm-prerequisites.html
https://aws.amazon.com/blogs/aws/amazon-ec2-now-supports-nitrotpm-and-uefi-secure-boot/

Comment (goneri) [reply tests/integration/targets/ec2_ami/tasks/main.yml:695]: This will be a good candidate for an unit-test.

**Review by goneri — APPROVED** [review]

File: tests/integration/targets/ec2_ami/tasks/main.yml (hunk 3 above)
Comment (mandar242) [reply tests/integration/targets/ec2_ami/tasks/main.yml:697]: Added a unit test, review would be helpful.
https://github.com/ansible-collections/amazon.aws/pull/1037/files#diff-e73e5db91a3de8db23f7c80c49a50b547462a582d68095d4580d2170c90b30f8

**Review by goneri — APPROVED** [review]

**Review by gravesm — APPROVED** [review]

**Pull request merged**